File: CS/InventoryChart_6/Form1.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Windows.Forms;
using DevExpress.XtraCharts;


namespace InventoryChart
{
    public partial class Form1 : Form
    {
        public Form1()
        {
            InitializeComponent();
        }
        

        private void Form1_Load(object sender, EventArgs e)
        {
            this.bOBindingSource.DataSource = new BO();
        }

        private void inventoryChart_ObjectSelected(object sender, DevExpress.XtraCharts.HotTrackEventArgs e)
        {
            if (e.HitInfo.InSeries && e.AdditionalObject is SeriesPoint)
            {
                SeriesPoint currentPoint = (SeriesPoint)e.AdditionalObject;
                if (currentPoint.Tag.GetType() == typeof(InventoryChart.BO.CategoryItem))
                {
                    InventoryChart.BO.CategoryItem currentCategory = (InventoryChart.BO.CategoryItem)currentPoint.Tag;
                    int categoryID = currentCategory.CategoryID;
                    DataFilter currentCategoryFilter = new DataFilter("ProductItems.CategoryID", "System.Int32", DataFilterCondition.Equal, categoryID);
                    this.inventoryChart.Series["ProductUnits"].DataFilters.Clear();
                    this.inventoryChart.Series["ProductUnits"].DataFilters.Add(currentCategoryFilter);
                    this.inventoryChart.Series["CategoryUnits"].Visible = false;
                    this.inventoryChart.Series["ProductUnits"].Visible = true;
                    this.inventoryChart.Legend.AlignmentHorizontal = LegendAlignmentHorizontal.RightOutside;
                    this.inventoryChart.Legend.AlignmentVertical = LegendAlignmentVertical.Center;
                    this.inventoryChart.Annotations[0].Visible = true;
                    PieSeriesView view = (PieSeriesView)this.inventoryChart.Series["ProductUnits"].View;
                    view.Titles[0].Text = currentCategory.CategoryName;
                }
            }
            else if (e.HitInfo.InAnnotation && e.Object is TextAnnotation)
            {
                this.inventoryChart.Annotations[0].Visible = false;
                this.inventoryChart.Series["CategoryUnits"].Visible = true;
                this.inventoryChart.Series["ProductUnits"].Visible = false;
                this.inventoryChart.Legend.AlignmentHorizontal = LegendAlignmentHorizontal.Center;
                this.inventoryChart.Legend.AlignmentVertical = LegendAlignmentVertical.TopOutside;
            }
        }

        private void inventoryChart_ObjectHotTracked(object sender, HotTrackEventArgs e)
        {
            if (e.HitInfo.InAnnotation && e.Object is TextAnnotation)
                Cursor.Current = Cursors.Hand;
            else Cursor.Current = Cursors.Default;
        }

        private void inventoryChart_CustomDrawCrosshair(object sender, CustomDrawCrosshairEventArgs e)
        {

            foreach (CrosshairElement element in e.CrosshairElements)
            {
                SeriesPoint currentPoint = element.SeriesPoint;
                if (currentPoint.Tag.GetType() == typeof(InventoryChart.BO.CategoryItem))
                {
                    InventoryChart.BO.CategoryItem currentCategory = (InventoryChart.BO.CategoryItem)currentPoint.Tag;
                    List<InventoryChart.BO.ProductItem> productItems = ((BO)this.bOBindingSource.DataSource).ProductItems;
                    InventoryChart.BO.ProductItem mostExpensiveProduct = productItems.Where(p => p.CategoryID == currentCategory.CategoryID).OrderByDescending(p => p.ProductPrice).First() as InventoryChart.BO.ProductItem;
                    element.LabelElement.Text += string.Format("\nMost expensive product:\n{0} : {1:c}",
                                                            mostExpensiveProduct.ProductName,
                                                            mostExpensiveProduct.ProductPrice);
                }
            }
        }


    }
}

File: VB/InventoryChart_6/Form1.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Linq
Imports System.Text
Imports System.Windows.Forms
Imports DevExpress.XtraCharts


Namespace InventoryChart
	Partial Public Class Form1
		Inherits Form
		Public Sub New()
			InitializeComponent()
		End Sub


		Private Sub Form1_Load(ByVal sender As Object, ByVal e As EventArgs) Handles MyBase.Load
			Me.bOBindingSource.DataSource = New BO()
		End Sub

		Private Sub inventoryChart_ObjectSelected(ByVal sender As Object, ByVal e As DevExpress.XtraCharts.HotTrackEventArgs) Handles inventoryChart.ObjectSelected
			If e.HitInfo.InSeries AndAlso TypeOf e.AdditionalObject Is SeriesPoint Then
				Dim currentPoint As SeriesPoint = CType(e.AdditionalObject, SeriesPoint)
				If currentPoint.Tag.GetType() Is GetType(InventoryChart.BO.CategoryItem) Then
					Dim currentCategory As InventoryChart.BO.CategoryItem = CType(currentPoint.Tag, InventoryChart.BO.CategoryItem)
					Dim categoryID As Integer = currentCategory.CategoryID
					Dim currentCategoryFilter As New DataFilter("ProductItems.CategoryID", "System.Int32", DataFilterCondition.Equal, categoryID)
					Me.inventoryChart.Series("ProductUnits").DataFilters.Clear()
					Me.inventoryChart.Series("ProductUnits").DataFilters.Add(currentCategoryFilter)
					Me.inventoryChart.Series("CategoryUnits").Visible = False
					Me.inventoryChart.Series("ProductUnits").Visible = True
					Me.inventoryChart.Legend.AlignmentHorizontal = LegendAlignmentHorizontal.RightOutside
					Me.inventoryChart.Legend.AlignmentVertical = LegendAlignmentVertical.Center
					Me.inventoryChart.Annotations(0).Visible = True
					Dim view As PieSeriesView = CType(Me.inventoryChart.Series("ProductUnits").View, PieSeriesView)
					view.Titles(0).Text = currentCategory.CategoryName
				End If
			ElseIf e.HitInfo.InAnnotation AndAlso TypeOf e.Object Is TextAnnotation Then
				Me.inventoryChart.Annotations(0).Visible = False
				Me.inventoryChart.Series("CategoryUnits").Visible = True
				Me.inventoryChart.Series("ProductUnits").Visible = False
				Me.inventoryChart.Legend.AlignmentHorizontal = LegendAlignmentHorizontal.Center
				Me.inventoryChart.Legend.AlignmentVertical = LegendAlignmentVertical.TopOutside
			End If
		End Sub

		Private Sub inventoryChart_ObjectHotTracked(ByVal sender As Object, ByVal e As HotTrackEventArgs) Handles inventoryChart.ObjectHotTracked
			If e.HitInfo.InAnnotation AndAlso TypeOf e.Object Is TextAnnotation Then
				Cursor.Current = Cursors.Hand
			Else
				Cursor.Current = Cursors.Default
			End If
		End Sub

		Private Sub inventoryChart_CustomDrawCrosshair(ByVal sender As Object, ByVal e As CustomDrawCrosshairEventArgs) Handles inventoryChart.CustomDrawCrosshair

			For Each element As CrosshairElement In e.CrosshairElements
				Dim currentPoint As SeriesPoint = element.SeriesPoint
				If currentPoint.Tag.GetType() Is GetType(InventoryChart.BO.CategoryItem) Then
					Dim currentCategory As InventoryChart.BO.CategoryItem = CType(currentPoint.Tag, InventoryChart.BO.CategoryItem)
					Dim productItems As List(Of InventoryChart.BO.ProductItem) = (CType(Me.bOBindingSource.DataSource, BO)).ProductItems
					Dim mostExpensiveProduct As InventoryChart.BO.ProductItem = TryCast(productItems.Where(Function(p) p.CategoryID = currentCategory.CategoryID).OrderByDescending(Function(p) p.ProductPrice).First(), InventoryChart.BO.ProductItem)
					element.LabelElement.Text += String.Format(Constants.vbLf & "Most expensive product:" & Constants.vbLf & "{0} : {1:c}", mostExpensiveProduct.ProductName, mostExpensiveProduct.ProductPrice)
				End If
			Next element
		End Sub


	End Class
End Namespace
